app/channel.go

```diff
@@ -954,55 +982,149 @@ func LeaveChannel(channelId string, userId string) *model.AppError {
+		if val, ok := post.Props["messages"]; ok {
+			post.Props["messages"] = append(
+				val.([]interface{}),
+				model.StringMap{"type": model.POST_ADD_TO_CHANNEL, "username": user.Username, "addedUsername": addedUser.Username},
+			)
+		} else {
+			oldProps := make(model.StringInterface)
+			for key, value := range post.Props {
+				oldProps[key] = value
+			}
+			oldProps["type"] = post.Type
+			post.Props["messages"] = []interface{}{
+				oldProps,
+				model.StringMap{"type": model.POST_ADD_TO_CHANNEL, "username": user.Username, "addedUsername": addedUser.Username},
+			}
+		}
+
+		if _, err := UpdatePost(post, false); err != nil {
+			return model.NewLocAppError("PostAddToChannelMessage", "api.channel.post_user_add_remove_message_and_forget.error", nil, err.Error())
+		}
+
+		SendNotificationsForSystemMessageAddRemove(user, addedUser, channel, post)
+	} else {
+		post := &model.Post{
+			ChannelId: channel.Id,
+			Message:   message,
+			Type:      model.POST_ADD_TO_CHANNEL,
+			UserId:    user.Id,
+			Props: model.StringInterface{
+				"username":      user.Username,
+				"addedUsername": addedUser.Username,
+			},
+		}
+
+		if _, err := CreatePost(post, channel.TeamId, false); err != nil {
+			return model.NewLocAppError("postAddToChannelMessage", "api.channel.post_user_add_remove_message_and_forget.error", nil, err.Error())
+		}
 	}
 
 	return nil
 }
 
 func PostRemoveFromChannelMessage(removerUserId string, removedUser *model.User, channel *model.Channel) *model.AppError {
-	post := &model.Post{
-		ChannelId: channel.Id,
-		Message:   fmt.Sprintf(utils.T("api.channel.remove_member.removed"), removedUser.Username),
-		Type:      model.POST_REMOVE_FROM_CHANNEL,
-		UserId:    removerUserId,
-		Props: model.StringInterface{
-			"removedUsername": removedUser.Username,
-		},
+	message := fmt.Sprintf(utils.T("api.channel.remove_member.removed"), removedUser.Username)
+
+	removerUser, err := GetUser(removerUserId)
+	if err != nil {
+		fmt.Printf("err: %+v\n", err)
+		return err
 	}
 
-	if _, err := CreatePost(post, channel.TeamId, false); err != nil {
-		return model.NewLocAppError("postRemoveFromChannelMessage", "api.channel.post_user_add_remove_message_and_forget.error", nil, err.Error())
+	post, err := GetLastPostForChannelIfJoinLeave(channel.Id)
+	if err == nil {
+		post.Message += " " + message
+
+		if val, ok := post.Props["messages"]; ok {
+			post.Props["messages"] = append(
+				val.([]interface{}),
+				model.StringMap{"type": model.POST_REMOVE_FROM_CHANNEL, "removedUsername": removedUser.Username},
+			)
+		} else {
+			oldProps := make(model.StringInterface)
+			for key, value := range post.Props {
+				oldProps[key] = value
+			}
+			oldProps["type"] = post.Type
+			post.Props["messages"] = []interface{}{
+				oldProps,
+				model.StringMap{"type": model.POST_REMOVE_FROM_CHANNEL, "removedUsername": removedUser.Username},
+			}
+		}
```

Comment:
You're repeating this `if/else` block four times with very few differences. Could we avoid some of this repetition by moving it into a function?